### PR TITLE
Importance Matrix calculation

### PR DESCRIPTION
**Pull request by ikawrakow**
### TL;DR

This PR adds to `llama.cpp` the ability to compute an "importance matrix" that can later be used for model quantization. The resulting matrix is much simpler and smaller compared to what is commonly used (see below for details), but, as far as I can tell, still leads to SOTA quantization results. 

**Caveat**
The tool only works **on the CPU**. As `llama.cpp` and `ggml` currently do not offer (easy) ways for extracting information about intermediate results of the computation graph, it would have been much more involved to add importance matrix collection to the CUDA/Metal back-ends. So, while I could have come up with something that works, I decided it is better to leave it to the project maintainers to implement what I'm sure will be a much more elegant solution. But because the calculation needs to run on the CPU, it does take some time. On my 32-core Ryzen-5975WX CPU one needs ~10 minutes to gather data for 50k tokens (the typical number of training tokens I have been using) for a 7B model, and about one hour for LLaMA-v2-70B.  

### Usage
```
./imatrix -m <some_fp_model> -f <some_training_data> [-o <output_file>] [--verbosity <verbosity_level>]
        [-ofreq num_chunks] [-ow <0 or 1>] [other GPT params]
```
Here `-m` with a model name and `-f` with a file containing training data (such as e.g. `wiki.train.raw`) are mandatory.
The parameters in square brackets are optional and have the following meaning:
* `-o` (or `--output-file`) specifies the name of the file where the computed data will be stored. If missing `imatrix.dat` is used.
* `--verbosity` specifies the verbosity level. If set to `0`, no output other than the perplexity of the processed chunks will be generated. If set to `1`, each time the results are saved a message is written to `stderr`. If `>=2`, a message is output each time data is collected for any tensor. Default verbosity level is `1`.
* `-ofreq` (or `--output-frequency`) specifies how often the so far computed result is saved to disk. Default is 10 (i.e., every 10 chunks)
* `-ow` (or `--output-weight`) specifies if data will be collected for the `output.weight` tensor. My experience is that it is better to not utilize the importance matrix when quantizing `output.weight`, so this is set to `false` by default.

`imatrix` being an example in `llama.cpp` understands all other GPT parameters such as `-t num_threads`, `--chunks num_chnks`, `-c context_length`, `-b batch_size`, etc., that can be relevant for the calculation.

### Remarks

The functionality could have been implemented in the existing `perplexity` example (and this is what I did in my private repo). However, I'm assuming that, once this is out in the wild, there will be a lot of experimentation. For instance, it currently just splits the provided training text into chunks of `n_ctx` tokens, and runs those through the evaluation pipeline. This is likely OK for general purpose models. But for instruct tuned models, it may actually be better to use instruct tuning training datasets for computing the importance matrix. If that turns out to be the case, it would be better to split the training data into the individual instruct entries rather than chunks of fixed length. This would require adding a new mode for performing the calculation. I think it is much cleaner and easier to add such additional capabilities to a dedicated tool instead of trying to stuff everything into `perplexity.cpp`. Hence the addition as a new example.  

### Details

The common approach for computing an importance matrix is to get the weights gradients from a training run on a given set of training tokens. When the gradient of a given model weight is small, it means that a large change in the model weight will result in a small change in model performance, and, vice versa, a large gradient implies a large change in model performance from a small change in model weight. Hence, one can use the squared gradients as an importance matrix. There are two problems with this:
* The training capabilities of `llama.cpp` are not quite there yet
* The importance matrix is huge (at least the same as the model size, or perhaps even double the size for `fp32` gradients)

Given this, I wanted an alternative approach. To get there, let's consider some row in a model tensor. This row will be multiplying the columns of intermediate results (a.k.a. "activations" or "embeddings") produced by network portions preceding the tensor in the computation graph. For simplicity let's consider a single "activation" column. The dot product of this tensor row with this activation column will result in one "activation" entry in the "activation" matrix passed down to subsequent parts of the computation graph. Let's call the tensor row entries `w_j` and the activation column entries `a_j`. Let's now assume that we will be quantizing this tensor row with resulting quantized weights `q_j`. To minimize the impact of the quantization we want to minimize, e.g., `F = [Sum_j (q_j - w_j) a_j]^2`. The derivative of `F` with respect to `q_i` is `Sum_j a_i a_j (q_j - w_j)`. This is for one "activation" column incident on the tensor row. If we average over a given set of training data, the above turns into `Sum_j <a_j a_j> (q_j - w_j)`, where `<x>` stands for the expectation value. Activations have positive and negative signs and so, unless there is some really strong correlation between activation `i` and activation `j`, we will have `<a_j a_j>` be much smaller than `<a_i^2>` for non-diagonal elements. Hence, we can greatly simplify things by computing just the diagonal elements `<a_i^2>`, which we can store consecutively in an array of size given by the hidden dimension of the tensor. We can then use these `<a_i^2>` as importances (weights) in a weighted RMSE minimization when quantizing the tensor. 

I have been using this approach in my private development repo and, given obtained SOTA results for 2-5 bit quantized models, I think it is fully adequate despite ignoring non-diagonal elements.

### CODE REVIEW

File: examples/imatrix/imatrix.cpp [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+    std::unordered_map<std::string, Stats> m_stats;
+    StatParams                             m_params;
+    std::mutex                             m_mutex;
+    int                                    m_last_call = 0;
+};
+
+void IMatrixCollector::collect_imatrix(const struct ggml_tensor * src0, const struct ggml_tensor * src1) {
+    if (src1->ne[1] < 16 || src1->type != GGML_TYPE_F32) return;
+    if (!(strncmp(src0->name, "blk.", 4) == 0 || (m_params.collect_output_weight && strcmp(src0->name, "output.weight") == 0))) return;
+    std::lock_guard<std::mutex> lock(m_mutex);
```
Comment (ggerganov) [inline examples/imatrix/imatrix.cpp:48]: This lock seems redundant because only `ith == 1` calls the callback

File: examples/imatrix/imatrix.cpp [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    gpt_params params;
+    params.n_batch = 512;
+    if (!gpt_params_parse(args.size(), args.data(), params)) {
+        return 1;
+    }
+
+    g_collector.set_parameters(std::move(sparams));
+
+    ggml_set_imatrix_collection(ik_collect_imatrix);
+    ggml_set_imatrix_collection(ik_collect_imatrix);
```
Comment (ggerganov) [inline examples/imatrix/imatrix.cpp:324]:
```suggestion
    ggml_set_imatrix_collection(ik_collect_imatrix);
```

**Review by ggerganov — APPROVED** [review]

**Comment by Ttl** [comment]
>The common approach for computing an importance matrix is to get the weights gradients from a training run on a given set of training tokens.

GPTQ uses the same importance matrix including non-diagonal elements: average of `a*a^T` over the calibration dataset for each layer. It equals Hessian of `F`.

I'm not aware of any popular post-training quantization methods that use gradients calculated with backpropagation.

**Comment by ikawrakow** [comment]
@Ttl 

> > The common approach for computing an importance matrix is to get the weights gradients from a training run on a given set of training tokens.
> 
> GPTQ uses the same importance matrix including non-diagonal elements: average of `a*a^T` over the calibration dataset for each layer. It equals Hessian of `F`.
> 
> I'm not aware of any popular post-training quantization methods that use gradients calculated with backpropagation.

I must have misunderstood, then. But the novelty claim does not come based on the usage `<a a^T>`, but rather the realization that one can drop non-diagonal elements in this matrix. Obviously we are dealing with completely different (much simpler!) minimization problems when just the diagonal elements are involved compared to the full `<a a^T>` (or the squared gradients from a training run).

**Comment by pipixia244** [comment]
About the quant data, do I need to change the text to the SFT format (e.g. for chatml, it is <|im_start|>user\nsomething<|im_end|>......) or just lines of raw text? (I'm using a chat model)

**Comment by ikawrakow** [comment]
> About the quant data, do I need to change the text to the SFT format (e.g. for chatml, it is <|im_start|>user\nsomething<|im_end|>......) or just lines of raw text? (I'm using a chat model)

I don't know the answer to this question. It will be a matter of experimentation. But my guess is that if one wanted to use a training dataset used for chat/instruct tuning that contains the control tokens, it might be better to split the training data into the individual training entries. This is not implemented (yet). The current implementation simply splits into chunks of `n_ctx` tokens and runs the inference. So that, if one used th SFT format, `<|im_start|>/<|im_end|>` will end up in random locations within a chunk. This may or may not be good.

**Comment by TheBloke** [comment]
@ikawrakow Amazing work!

Stupid question:  "some fp model" is the model to be quantised, I assume?  Ie this needs to be done uniquely for each model to be quantised, it can't be done once and re-used for future models of the same arch/base model/tokenizer/anything?

**Comment by ikawrakow** [comment]
@TheBloke With PR #4957 the importance matrix calculation can be done on the GPU, which makes it way faster, so running this calculation for every model to be quantized will be no longer such a pain as it was for me when I was developing the stuff.

Note that I'm preparing a PR to add importance matrix support for `Q4_0, Q4_1, Q5_0, Q5_1`, so perhaps it makes sense to wait for this to be also merged before you embark on publishing the new quantized models.

**Comment by pipixia244** [comment]
> > About the quant data, do I need to change the text to the SFT format (e.g. for chatml, it is <|im_start|>user\nsomething<|im_end|>......) or just lines of raw text? (I'm using a chat model)
> 
> I don't know the answer to this question. It will be a matter of experimentation. But my guess is that if one wanted to use a training dataset used for chat/instruct tuning that contains the control tokens, it might be better to split the training data into the individual training entries. This is not implemented (yet). The current implementation simply splits into chunks of `n_ctx` tokens and runs the inference. So that, if one used the SFT format, `<|im_start|>/<|im_end|>` will end up in random locations within a chunk. This may or may not be good.

@ikawrakow Thank you for your great work and your response! Looking forward to your future work on individual training entries for chat data.

Also another question about the data, can this quant method works with masks like SFT? When I'm using chat data, I only want the model to predict the assistant part of the whole chat data (for example, only calculate "yyyyyy" below: <|im_start|>user\n xxxxxx<|im_end|>\n<|im_start|>assistant\n yyyyyyyyyy<|im_end|>\n).

**Comment by sorasoras** [comment]
> @TheBloke With PR #4957 the importance matrix calculation can be done on the GPU, which makes it way faster, so running this calculation for every model to be quantized will be no longer such a pain as it was for me when I was developing the stuff.
> 
> Note that I'm preparing a PR to add importance matrix support for `Q4_0, Q4_1, Q5_0, Q5_1`, so perhaps it makes sense to wait for this to be also merged before you embark on publishing the new quantized models.

Same as above.
I wonder if it is possible to just calculate important matrix against response instead of prompt.

**Comment by DavidZyy** [comment]
I write a document about imatrix in my understanding with the help of gpt [here](https://github.com/DavidZyy/algorithm_notes/blob/main/LLM/imatrix.md)，but it seems that github not support latex math formula well, you can download the `.md`.
